src/Geometry/CartesianDiscreteModels.jl

```diff
@@ -199,3 +236,132 @@ function _fix_dface_geolabels!(
     end
   end
 end
+
+function _fill_subgrid_cartesian_face_labeling!(labels,topo,subdesc,desc,cmin)
+   _fill_subgrid_cartesian_entities!(labels,topo,subdesc,desc,cmin)
+   _add_cartesian_tags!(labels,topo)
+end
+
+function _fill_subgrid_cartesian_entities!(labels, topo, subdesc, desc, cmin)
+  D = num_cell_dims(topo)
+  d_to_dface_to_entity = labels.d_to_dface_to_entity
+  gcis = CartesianIndices(Tuple(desc.partition))
+  subcis = CartesianIndices(Tuple(subdesc.partition))
+
+  polytope = first(get_polytopes(topo))
+  face_labeling = labels
+  offsets = get_offsets(polytope)
+  interior_id = num_faces(polytope)
+  boundary_id = -1
+
+  minus_one_ci = CartesianIndex(tfill(-1, Val{D}()))
+
+  polytope_d_face_to_jfaces = Matrix{Vector{Vector{Int}}}(undef, (D, D))
+  for d = 0:(D-1)
+    for j = d+1:D-1
+      polytope_d_face_to_jfaces[d+1, j+1] = get_faces(polytope, d, j)
+    end
+  end
+
+  face_deltas = _find_ncube_face_neighbor_deltas(polytope)
+  for d = 0:(D-1)
+    face_to_cells = get_faces(topo, d, D)
+    cell_to_faces = get_faces(topo, D, d)
+    face_to_geolabel = face_labeling.d_to_dface_to_entity[d+1]
+    nfaces = length(face_to_geolabel)
+    for face_gid = 1:nfaces
```

Comment:
@amartinhuertas Could you put long loops like this one (i.e. loops over cells or over faces) inside a [function barrier](https://docs.julialang.org/en/v1/manual/performance-tips/index.html#kernel-functions-1)?
This is a way to ensure type stability in computationally critical parts

Reply:
Just to double-check, you are asking that I move the loop in lines 272-315 to a separate function, right? I will definitely do it.

On the other hand, while I clearly understand the issue in the Julia Manual's example, I fail to see why here there is a type stability issue (I did not check it with @code_typed yet, I will do it). Where does the type instability stems from? `face_to_cells`, `cell_to_faces` ? anything else?

Reply:
The problem is that some lines at the begging of the function 
```julia
function _fill_subgrid_cartesian_entities!(labels, topo, subdesc, desc, cmin)
```
are potentially type unstable, like

```julia
polytope = first(get_polytopes(topo))
```

For performance reasons, a common pattern in Gridap is to split functions into a small type unstable pre-process and a type-stable computationally expensive part.

Reply:
> The problem is that some lines at the beginning of the function
function _fill_subgrid_cartesian_entities!(labels, topo, subdesc, desc, cmin)
are potentially type unstable, like
polytope = first(get_polytopes(topo))

Ok. I see now the point after using the  `@code_warntype` macro:

```

julia> @code_warntype Gridap.Geometry._fill_subgrid_cartesian_entities!(model.face_labeling, model.grid_topology, desc, desc, CartesianIndex(2,2,2))
Variables
  #self#::Core.Compiler.Const(Gridap.Geometry._fill_subgrid_cartesian_entities!, false)
  labels::Gridap.Geometry.FaceLabeling
  topo::Gridap.Geometry.UnstructuredGridTopology{3,3,Float64,true}
  subdesc::Gridap.Geometry.CartesianDescriptor{3,Float64,typeof(identity)}
  desc::Gridap.Geometry.CartesianDescriptor{3,Float64,typeof(identity)}
  cmin::CartesianIndex{3}
  D::Int64
  d_to_dface_to_entity::Array{Array{Int32,1},1}
  gcis::CartesianIndices{3,Tuple{Base.OneTo{Int64},Base.OneTo{Int64},Base.OneTo{Int64}}}
  subcis::CartesianIndices{3,Tuple{Base.OneTo{Int64},Base.OneTo{Int64},Base.OneTo{Int64}}}
  polytope::Any
  face_labeling::Gridap.Geometry.FaceLabeling
  offsets::Any
  interior_id::Any
  boundary_id::Int64
  minus_one_ci::CartesianIndex{3}
  polytope_d_face_to_jfaces::Array{Array{Array{Int64,1},1},2}
  @_18::Union{Nothing, Tuple{Int64,Int64}}
  face_deltas::Array{CartesianIndex,1}
  @_20::Union{Nothing, Tuple{Int64,Int64}}
  d@_21::Int64
  @_22::Union{Nothing, Tuple{Int64,Int64}}
  j@_23::Int64
  face_to_cells::Gridap.Arrays.Table{Int64,Int32}
  cell_to_faces::Gridap.Arrays.Table{Int64,Int32}
  face_to_geolabel::Array{Int32,1}
  nfaces::Int64
  @_28::Union{Nothing, Tuple{Int64,Int64}}
  d@_29::Int64
  face_gid::Int64
  cell_gid::Int64
  a::Int32
  b::Int64
  gci::CartesianIndex{3}
  @_35::Union{Nothing, Tuple{Int64,Int64}}
  face_lid::Any
  is_assigned_face_delta::Bool
  cell_found::Bool
  @_39::Union{Nothing, Tuple{Int64,Int64}}
  j@_40::Int64
  dface_to_jfaces::Any
  j@_42::Int64
  @_43::Union{Missing, Bool}

Body::Array{Int32,1}
1 ──        Core.NewvarNode(:(face_deltas))
│           Core.NewvarNode(:(@_20))
│           (D = Gridap.Geometry.num_cell_dims(topo))
│           (d_to_dface_to_entity = Base.getproperty(labels, :d_to_dface_to_entity))
│    %5   = Base.getproperty(desc, :partition)::Gridap.TensorValues.MultiValue{Tuple{3},Int64,1,3}
│    %6   = Gridap.Geometry.Tuple(%5)::Tuple{Int64,Int64,Int64}
│           (gcis = Gridap.Geometry.CartesianIndices(%6))
│    %8   = Base.getproperty(subdesc, :partition)::Gridap.TensorValues.MultiValue{Tuple{3},Int64,1,3}
│    %9   = Gridap.Geometry.Tuple(%8)::Tuple{Int64,Int64,Int64}
│           (subcis = Gridap.Geometry.CartesianIndices(%9))
│    %11  = Gridap.Geometry.get_polytopes(topo)::Array{_A,1} where _A
│           (polytope = Gridap.Geometry.first(%11))
│           (face_labeling = labels)
│           (offsets = Gridap.Geometry.get_offsets(polytope))
│           (interior_id = Gridap.Geometry.num_faces(polytope))
│           (boundary_id = -1)
│    %17  = Core.apply_type(Gridap.Geometry.Val, D::Core.Compiler.Const(3, false))::Core.Compiler.Const(Val{3}, false)
│    %18  = (%17)()::Core.Compiler.Const(Val{3}(), false)
│    %19  = Gridap.Geometry.tfill(-1, %18)::Core.Compiler.Const((-1, -1, -1), false)
│           (minus_one_ci = Gridap.Geometry.CartesianIndex(%19))
│    %21  = Core.apply_type(Gridap.Geometry.Vector, Gridap.Geometry.Int)::Core.Compiler.Const(Array{Int64,1}, false)
│    %22  = Core.apply_type(Gridap.Geometry.Vector, %21)::Core.Compiler.Const(Array{Array{Int64,1},1}, false)
│    %23  = Core.apply_type(Gridap.Geometry.Matrix, %22)::Core.Compiler.Const(Array{Array{Array{Int64,1},1},2}, false)
│    %24  = Core.tuple(D::Core.Compiler.Const(3, false), D::Core.Compiler.Const(3, false))::Core.Compiler.Const((3, 3), false)
│           (polytope_d_face_to_jfaces = (%23)(Gridap.Geometry.undef, %24))
│    %26  = (D::Core.Compiler.Const(3, false) - 1)::Core.Compiler.Const(2, false)
│    %27  = (0:%26)::Core.Compiler.Const(0:2, false)
│           (@_18 = Base.iterate(%27))
│    %29  = (@_18::Core.Compiler.Const((0, 0), false) === nothing)::Core.Compiler.Const(false, false)
│    %30  = Base.not_int(%29)::Core.Compiler.Const(true, false)
└───        goto #7 if not %30
2 ┄─ %32  = @_18::Tuple{Int64,Int64}::Tuple{Int64,Int64}
│           (d@_21 = Core.getfield(%32, 1))
│    %34  = Core.getfield(%32, 2)::Int64
│    %35  = (d@_21 + 1)::Int64
│    %36  = (D::Core.Compiler.Const(3, false) - 1)::Core.Compiler.Const(2, false)
│    %37  = (%35:%36)::UnitRange{Int64}
│           (@_22 = Base.iterate(%37))
│    %39  = (@_22 === nothing)::Bool
│    %40  = Base.not_int(%39)::Bool
└───        goto #5 if not %40
3 ┄─ %42  = @_22::Tuple{Int64,Int64}::Tuple{Int64,Int64}
│           (j@_23 = Core.getfield(%42, 1))
│    %44  = Core.getfield(%42, 2)::Int64
│    %45  = Gridap.Geometry.get_faces(polytope, d@_21, j@_23)::Any
│    %46  = polytope_d_face_to_jfaces::Array{Array{Array{Int64,1},1},2}
│    %47  = (d@_21 + 1)::Int64
│    %48  = (j@_23 + 1)::Int64
│           Base.setindex!(%46, %45, %47, %48)
│           (@_22 = Base.iterate(%37, %44))
│    %51  = (@_22 === nothing)::Bool
│    %52  = Base.not_int(%51)::Bool
└───        goto #5 if not %52
4 ──        goto #3
5 ┄─        (@_18 = Base.iterate(%27, %34))
│    %56  = (@_18 === nothing)::Bool
│    %57  = Base.not_int(%56)::Bool
└───        goto #7 if not %57
6 ──        goto #2
7 ┄─        (face_deltas = Gridap.Geometry._find_ncube_face_neighbor_deltas(polytope))
│    %61  = (D::Core.Compiler.Const(3, false) - 1)::Core.Compiler.Const(2, false)
│    %62  = (0:%61)::Core.Compiler.Const(0:2, false)
│           (@_20 = Base.iterate(%62))
│    %64  = (@_20::Core.Compiler.Const((0, 0), false) === nothing)::Core.Compiler.Const(false, false)
│    %65  = Base.not_int(%64)::Core.Compiler.Const(true, false)
└───        goto #35 if not %65
8 ┄─ %67  = @_20::Tuple{Int64,Int64}::Tuple{Int64,Int64}
│           (d@_29 = Core.getfield(%67, 1))
│    %69  = Core.getfield(%67, 2)::Int64
│           (face_to_cells = Gridap.Geometry.get_faces(topo, d@_29, D::Core.Compiler.Const(3, false)))
│           (cell_to_faces = Gridap.Geometry.get_faces(topo, D::Core.Compiler.Const(3, false), d@_29))
│    %72  = Base.getproperty(face_labeling, :d_to_dface_to_entity)::Array{Array{Int32,1},1}
│    %73  = (d@_29 + 1)::Int64
│           (face_to_geolabel = Base.getindex(%72, %73))
│           (nfaces = Gridap.Geometry.length(face_to_geolabel))
│    %76  = (1:nfaces)::Core.Compiler.PartialStruct(UnitRange{Int64}, Any[Core.Compiler.Const(1, false), Int64])
│           (@_28 = Base.iterate(%76))
│    %78  = (@_28 === nothing)::Bool
│    %79  = Base.not_int(%78)::Bool
└───        goto #33 if not %79
9 ┄─        Core.NewvarNode(:(is_assigned_face_delta))
│           Core.NewvarNode(:(cell_found))
│           Core.NewvarNode(:(@_39))
│    %84  = @_28::Tuple{Int64,Int64}::Tuple{Int64,Int64}
│           (face_gid = Core.getfield(%84, 1))
│    %86  = Core.getfield(%84, 2)::Int64
│    %87  = Base.getproperty(face_to_cells, :data)::Array{Int64,1}
│    %88  = Base.getproperty(face_to_cells, :ptrs)::Array{Int32,1}
│    %89  = Base.getindex(%88, face_gid)::Int32
│           (cell_gid = Base.getindex(%87, %89))
│    %91  = Base.getproperty(cell_to_faces, :ptrs)::Array{Int32,1}
│           (a = Base.getindex(%91, cell_gid))
│    %93  = Base.getproperty(cell_to_faces, :ptrs)::Array{Int32,1}
│    %94  = (cell_gid + 1)::Int64
│    %95  = Base.getindex(%93, %94)::Int32
│           (b = %95 - 1)
│    %97  = Base.getindex(subcis, cell_gid)::CartesianIndex{3}
│    %98  = (%97 + minus_one_ci::Core.Compiler.Const(CartesianIndex(-1, -1, -1), false))::CartesianIndex{3}
│           (gci = %98 + cmin)
│           (face_lid = -1)
│    %101 = (a:b)::UnitRange{Int64}
│           (@_35 = Base.iterate(%101))
│    %103 = (@_35 === nothing)::Bool
│    %104 = Base.not_int(%103)::Bool
└───        goto #14 if not %104
10 ┄ %106 = @_35::Tuple{Int64,Int64}::Tuple{Int64,Int64}
│           (j@_40 = Core.getfield(%106, 1))
│    %108 = Core.getfield(%106, 2)::Int64
│    %109 = Base.getproperty(cell_to_faces, :data)::Array{Int64,1}
│    %110 = Base.getindex(%109, j@_40)::Int64
│    %111 = (%110 == face_gid)::Bool
└───        goto #12 if not %111
11 ─ %113 = (j@_40 - a)::Int64
│           (face_lid = %113 + 1)
└───        goto #14
12 ─        (@_35 = Base.iterate(%101, %108))
│    %117 = (@_35 === nothing)::Bool
│    %118 = Base.not_int(%117)::Bool
└───        goto #14 if not %118
13 ─        goto #10
14 ┄ %121 = (face_lid::Int64 != -1)::Bool
└───        goto #16 if not %121
15 ─        goto #17
16 ─ %124 = Base.AssertionError("face_lid != -1")::AssertionError
└───        Base.throw(%124)
17 ┄ %126 = face_lid::Int64::Int64
│    %127 = offsets::Any
│    %128 = (d@_29 + 1)::Int64
│    %129 = Base.getindex(%127, %128)::Any
│           (face_lid = %126 + %129)
│           (is_assigned_face_delta = Gridap.Geometry.isassigned(face_deltas, face_lid))
└───        goto #19 if not is_assigned_face_delta
18 ─ %133 = gci::CartesianIndex{3}
│    %134 = Base.getindex(face_deltas, face_lid)::Any
│    %135 = (%133 + %134)::Any
│           (@_43 = %135 in gcis)
└───        goto #20
19 ─        (@_43 = false)
20 ┄        goto #22 if not @_43
21 ─        Base.setindex!(face_to_geolabel, interior_id, face_gid)
└───        goto #31
22 ─        (cell_found = false)
│    %143 = (d@_29 + 1)::Int64
│    %144 = (D::Core.Compiler.Const(3, false) - 1)::Core.Compiler.Const(2, false)
│    %145 = (%143:%144)::UnitRange{Int64}
│           (@_39 = Base.iterate(%145))
│    %147 = (@_39 === nothing)::Bool
│    %148 = Base.not_int(%147)::Bool
└───        goto #28 if not %148
23 ┄ %150 = @_39::Tuple{Int64,Int64}::Tuple{Int64,Int64}
│           (j@_42 = Core.getfield(%150, 1))
│    %152 = Core.getfield(%150, 2)::Int64
│    %153 = polytope_d_face_to_jfaces::Array{Array{Array{Int64,1},1},2}
│    %154 = (d@_29 + 1)::Int64
│    %155 = (j@_42 + 1)::Int64
│    %156 = Base.getindex(%153, %154, %155)::Array{Array{Int64,1},1}
│    %157 = face_lid::Any
│    %158 = offsets::Any
│    %159 = (d@_29 + 1)::Int64
│    %160 = Base.getindex(%158, %159)::Any
│    %161 = (%157 - %160)::Any
│           (dface_to_jfaces = Base.getindex(%156, %161))
│    %163 = dface_to_jfaces::Any
│    %164 = offsets::Any
│    %165 = (j@_42 + 1)::Int64
│    %166 = Base.getindex(%164, %165)::Any
│    %167 = gcis::CartesianIndices{3,Tuple{Base.OneTo{Int64},Base.OneTo{Int64},Base.OneTo{Int64}}}
│    %168 = gci::CartesianIndex{3}
│           (cell_found = Gridap.Geometry._is_there_interior_cell_across_higher_dim_faces(%163, %166, %167, %168, face_deltas))
└───        goto #26 if not cell_found
24 ─        goto #28
25 ─        Core.Compiler.Const(:(goto %173), false)
26 ┄        (@_39 = Base.iterate(%145, %152))
│    %174 = (@_39 === nothing)::Bool
│    %175 = Base.not_int(%174)::Bool
└───        goto #28 if not %175
27 ─        goto #23
28 ┄        goto #30 if not cell_found
29 ─        Base.setindex!(face_to_geolabel, boundary_id::Core.Compiler.Const(-1, false), face_gid)
└───        goto #31
30 ─        Base.setindex!(face_to_geolabel, face_lid, face_gid)
31 ┄        (@_28 = Base.iterate(%76, %86))
│    %183 = (@_28 === nothing)::Bool
│    %184 = Base.not_int(%183)::Bool
└───        goto #33 if not %184
32 ─        goto #9
33 ┄        (@_20 = Base.iterate(%62, %69))
│    %188 = (@_20 === nothing)::Bool
│    %189 = Base.not_int(%188)::Bool
└───        goto #35 if not %189
34 ─        goto #8
35 ┄        Gridap.Geometry._fix_geolabels(D::Core.Compiler.Const(3, false), topo, d_to_dface_to_entity, interior_id, boundary_id::Core.Compiler.Const(-1, false))
│    %193 = d_to_dface_to_entity::Array{Array{Int32,1},1}
│    %194 = Base.lastindex(d_to_dface_to_entity)::Int64
│    %195 = Base.getindex(%193, %194)::Array{Int32,1}
│    %196 = Gridap.Geometry.fill!(%195, interior_id)::Array{Int32,1}
└───        return %196
```

Reply:
Hopefully the function barrier is type stable. Is it possible to check this with the debugger? 
Does this work? moving to just before calling the barrier, and call `@call_warntype ...` in the debugger prompt.

Reply:
> Is it possible to check this with the debugger?

Good point! I was wondering how to check this.  I have not been able to get `@code_warntype` going down in the call tree. Do u know if this is possible?

Does this work? moving to just before calling the barrier, and call @call_warntype ... in the debugger prompt.

Yes, I could check it with the debugger, but as always, I had to fight quite a lot. With the debugger stopped at the line at hand, I had to execute the following commands on the debugger prompt:

```
debug> using InteractiveUtils
┌ Warning: Package Gridap does not have InteractiveUtils in its dependencies:
│ - If you have Gridap checked out for development and have
│   added InteractiveUtils as a dependency but haven't updated your primary
│   environment's manifest file, try `Pkg.resolve()`.
│ - Otherwise you may need to report an issue with Gridap
└ Loading InteractiveUtils into Gridap from project dependency, future warnings for Gridap are suppressed.

debug> InteractiveUtils.@code_warntype _generate_subgrid_pregeo_label!(      face_to_geolabel,       d,      D,       offsets,      polytope_d_face_to_jfaces,       minus_one_ci,      face_to_cells,       cell_to_faces,      subcis,       gcis,      face_deltas,       cmin,      interior_id,       boundary_id,     )
```
with the output below. Now there is no "Any" in the output generated. But still the following line "%59  = Base.getindex(face_deltas, face_lid)::CartesianIndex" is shown with "CartesianIndex" in red. Any clue?

```
Variables
  #self#::Core.Compiler.Const(Gridap.Geometry._generate_subgrid_pregeo_label!, false)
  face_to_geolabel::Array{Int32,1}
  d::Int64
  D::Int64
  offsets::Array{Int64,1}
  polytope_d_face_to_jfaces::Array{Array{Array{Int64,1},1},2}
  minus_one_ci::CartesianIndex{3}
  face_to_cells::Gridap.Arrays.Table{Int64,Int32}
  cell_to_faces::Gridap.Arrays.Table{Int64,Int32}
  subcis::CartesianIndices{3,Tuple{Base.OneTo{Int64},Base.OneTo{Int64},Base.OneTo{Int64}}}
  gcis::CartesianIndices{3,Tuple{Base.OneTo{Int64},Base.OneTo{Int64},Base.OneTo{Int64}}}
  face_deltas::Array{CartesianIndex,1}
  cmin::CartesianIndex{3}
  interior_id::Int64
  boundary_id::Int64
  @_16::Union{Nothing, Tuple{Int64,Int64}}
  face_gid::Int64
  cell_gid::Int64
  a::Int32
  b::Int64
  gci::CartesianIndex{3}
  @_22::Union{Nothing, Tuple{Int64,Int64}}
  face_lid::Int64
  is_assigned_face_delta::Bool
  cell_found::Bool
  @_26::Union{Nothing, Tuple{Int64,Int64}}
  j@_27::Int64
  dface_to_jfaces::Array{Int64,1}
  j@_29::Int64
  @_30::Bool

Body::Nothing
1 ── %1   = Gridap.Geometry.length(face_to_geolabel)::Int64
│    %2   = (1:%1)::Core.Compiler.PartialStruct(UnitRange{Int64}, Any[Core.Compiler.Const(1, false), Int64])
│           (@_16 = Base.iterate(%2))
│    %4   = (@_16 === nothing)::Bool
│    %5   = Base.not_int(%4)::Bool
└───        goto #26 if not %5
2 ┄─        Core.NewvarNode(:(is_assigned_face_delta))
│           Core.NewvarNode(:(cell_found))
│           Core.NewvarNode(:(@_26))
│    %10  = @_16::Tuple{Int64,Int64}::Tuple{Int64,Int64}
│           (face_gid = Core.getfield(%10, 1))
│    %12  = Core.getfield(%10, 2)::Int64
│    %13  = Base.getproperty(face_to_cells, :data)::Array{Int64,1}
│    %14  = Base.getproperty(face_to_cells, :ptrs)::Array{Int32,1}
│    %15  = Base.getindex(%14, face_gid)::Int32
│           (cell_gid = Base.getindex(%13, %15))
│    %17  = Base.getproperty(cell_to_faces, :ptrs)::Array{Int32,1}
│           (a = Base.getindex(%17, cell_gid))
│    %19  = Base.getproperty(cell_to_faces, :ptrs)::Array{Int32,1}
│    %20  = (cell_gid + 1)::Int64
│    %21  = Base.getindex(%19, %20)::Int32
│           (b = %21 - 1)
│    %23  = Base.getindex(subcis, cell_gid)::CartesianIndex{3}
│    %24  = (%23 + minus_one_ci)::CartesianIndex{3}
│           (gci = %24 + cmin)
│           (face_lid = -1)
│    %27  = (a:b)::UnitRange{Int64}
│           (@_22 = Base.iterate(%27))
│    %29  = (@_22 === nothing)::Bool
│    %30  = Base.not_int(%29)::Bool
└───        goto #7 if not %30
3 ┄─ %32  = @_22::Tuple{Int64,Int64}::Tuple{Int64,Int64}
│           (j@_27 = Core.getfield(%32, 1))
│    %34  = Core.getfield(%32, 2)::Int64
│    %35  = Base.getproperty(cell_to_faces, :data)::Array{Int64,1}
│    %36  = Base.getindex(%35, j@_27)::Int64
│    %37  = (%36 == face_gid)::Bool
└───        goto #5 if not %37
4 ── %39  = (j@_27 - a)::Int64
│           (face_lid = %39 + 1)
└───        goto #7
5 ──        (@_22 = Base.iterate(%27, %34))
│    %43  = (@_22 === nothing)::Bool
│    %44  = Base.not_int(%43)::Bool
└───        goto #7 if not %44
6 ──        goto #3
7 ┄─ %47  = (face_lid != -1)::Bool
└───        goto #9 if not %47
8 ──        goto #10
9 ── %50  = Base.AssertionError("face_lid != -1")::AssertionError
└───        Base.throw(%50)
10 ┄ %52  = face_lid::Int64
│    %53  = (d + 1)::Int64
│    %54  = Base.getindex(offsets, %53)::Int64
│           (face_lid = %52 + %54)
│           (is_assigned_face_delta = Gridap.Geometry.isassigned(face_deltas, face_lid))
└───        goto #12 if not is_assigned_face_delta
11 ─ %58  = gci::CartesianIndex{3}
│    %59  = Base.getindex(face_deltas, face_lid)::CartesianIndex
│    %60  = (%58 + %59)::CartesianIndex{3}
│           (@_30 = %60 in gcis)
└───        goto #13
12 ─        (@_30 = false)
13 ┄        goto #15 if not @_30
14 ─        Base.setindex!(face_to_geolabel, interior_id, face_gid)
└───        goto #24
15 ─        (cell_found = false)
│    %68  = (d + 1)::Int64
│    %69  = (D - 1)::Int64
│    %70  = (%68:%69)::UnitRange{Int64}
│           (@_26 = Base.iterate(%70))
│    %72  = (@_26 === nothing)::Bool
│    %73  = Base.not_int(%72)::Bool
└───        goto #21 if not %73
16 ┄ %75  = @_26::Tuple{Int64,Int64}::Tuple{Int64,Int64}
│           (j@_29 = Core.getfield(%75, 1))
│    %77  = Core.getfield(%75, 2)::Int64
│    %78  = (d + 1)::Int64
│    %79  = (j@_29 + 1)::Int64
│    %80  = Base.getindex(polytope_d_face_to_jfaces, %78, %79)::Array{Array{Int64,1},1}
│    %81  = face_lid::Int64
│    %82  = (d + 1)::Int64
│    %83  = Base.getindex(offsets, %82)::Int64
│    %84  = (%81 - %83)::Int64
│           (dface_to_jfaces = Base.getindex(%80, %84))
│    %86  = dface_to_jfaces::Array{Int64,1}
│    %87  = (j@_29 + 1)::Int64
│    %88  = Base.getindex(offsets, %87)::Int64
│    %89  = gci::CartesianIndex{3}
│           (cell_found = Gridap.Geometry._is_there_interior_cell_across_higher_dim_faces(%86, %88, gcis, %89, face_deltas))
└───        goto #19 if not cell_found
17 ─        goto #21
18 ─        Core.Compiler.Const(:(goto %94), false)
19 ┄        (@_26 = Base.iterate(%70, %77))
│    %95  = (@_26 === nothing)::Bool
│    %96  = Base.not_int(%95)::Bool
└───        goto #21 if not %96
20 ─        goto #16
21 ┄        goto #23 if not cell_found
22 ─        Base.setindex!(face_to_geolabel, boundary_id, face_gid)
└───        goto #24
23 ─        Base.setindex!(face_to_geolabel, face_lid, face_gid)
24 ┄        (@_16 = Base.iterate(%2, %12))
│    %104 = (@_16 === nothing)::Bool
│    %105 = Base.not_int(%104)::Bool
└───        goto #26 if not %105
25 ─        goto #2
26 ┄        return
```

Reply:
Great! This function looks green now! (except for the unions with nothing that are inevitable and unproblematic)

Reply:
@fverdugo ... I already finished addressing your third round of comments.

Reply:
> But still the following line "%59 = Base.getindex(face_deltas, face_lid)::CartesianIndex" is shown with "CartesianIndex" in red. Any clue?

This means that the compiler was not able to determine the `D` parameter of `CartesianIndex{D}`

It looks like you are mixing different values of `D` in `face_deltas`

Reply:
> @fverdugo ... I already finished addressing your third round of comments.

There is still room for a 4th round 😄 if you want to fix the `CartesianIndex` in red!

Reply:
> There is still room for a 4th round smile if you want to fix the CartesianIndex in red!

Solved in :metal: aa9e320e46dd57fe743196febd4cee571cf50eba

```
(d + 1)::Int64
│    %53  = Base.getindex(offsets, %52)::Int64
│           (face_lid = %51 + %53)
│    %55  = gci::CartesianIndex{3}
│    %56  = Base.getindex(face_deltas, face_lid)::CartesianIndex{3}
│    %57  = (%55 + %56)::CartesianIndex{3}
│    %58  = (%57 in gcis)::Bool
└───        goto #12 if not %58
```